### PR TITLE
Feature/arc 604 implement ephemeral deploy step

### PR DESCRIPTION
**Pull request by Jorg88**

### CODE REVIEW

File: src/mpyl/steps/deploy/ephemeraldockerdeploy.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+            description='Deploy ephemeral container',
+            version='0.0.1',
+            stage=Stage.DEPLOY
+        ), produced_artifact=ArtifactType.NONE, required_artifact=ArtifactType.DOCKER_IMAGE)
+
+    def execute(self, step_input: Input) -> Output:
+        env_file_name = write_env_to_file(step_input.project, step_input.run_properties)
+        docker_image_name = docker_image_tag(step_input)
+
+        return custom_check_output(self._logger,
```
Comment (koeves) [inline src/mpyl/steps/deploy/ephemeraldockerdeploy.py:26]: Maybe we can move this to the config file as well like with the sbt command

File: src/mpyl/utilities/docker/__init__.py [hunk 2]
```diff
@@ -92,3 +92,21 @@ def build(logger: Logger, root_path: str, file_path: str, image_tag: str, target
     stream_docker_logging(logger=logger, generator=logs, task_name=f'Build {file_path}:{target}')
     logger.debug(logs)
     return True
+
+
+def write_env_to_file(project: Project, run_properties: RunProperties) -> str:
```
Comment (koeves) [inline src/mpyl/utilities/docker/__init__.py:97]: Wondering if this should be here or in `mpyl.project.Properties`

File: src/mpyl/steps/deploy/ephemeraldockerdeploy.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+                                   f"docker run --rm --env-file ${env_file_name} ${docker_image_name}")
```
[the remark this answers is not in the document]
Comment (Jorg88) [reply src/mpyl/steps/deploy/ephemeraldockerdeploy.py]: You mean the `to_command` function? Isn't that a bit much for such a simple cmd that is only used here?

File: src/mpyl/utilities/docker/__init__.py (hunk 2 above)
Comment (Jorg88) [reply src/mpyl/utilities/docker/__init__.py:97]: You tell me, I have no idea! ;)

File: src/mpyl/steps/deploy/ephemeraldockerdeploy.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (koeves) [reply src/mpyl/steps/deploy/ephemeraldockerdeploy.py]: No more like `https://github.com/Vandebron/mpyl/blob/bb44d5d075cf4fda781b81d875ee5f0aef297077/config.yml#L42-L50`

[the remark this answers is not in the document]
Comment (Jorg88) [reply src/mpyl/steps/deploy/ephemeraldockerdeploy.py]: Yeah and that is used in `src/mpyl/utilities/sbt/__init__.py`. If you ask me it seems like a bit of overkill to create a complete class with props and a function that builds the command for something so simple that is not re-used (afaik). But maybe I've been working in a functional front-end programming environment for too long, hehe.

Let me know, I'll do it if you and Sam think it's better!

File: tests/steps/deploy/ephemeral/test_ephemeral.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class TestEphemeral:
+    resource_path = root_test_path / 'projects' / 'ephemeral' / 'deployment'
+
+    def test_replace_default_with_none(self):
+        step_input = Input(load_project(self.resource_path, Path('project.yml'), True), test_data.RUN_PROPERTIES, None)
+        assert len(step_input.project.deployment.properties.env) == 4
+        write_env_to_file(step_input.project, test_data.RUN_PROPERTIES)
+        env_file_name = step_input.run_properties.config['docker']['envFileName']
```
Comment (SamTheisens) [inline tests/steps/deploy/ephemeral/test_ephemeral.py:18]: I think you could use a temp file for this, that takes care of removal by itsself

File: tests/projects/ephemeral/deployment/Dockerfile-mpl [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# syntax=docker/dockerfile:1.4
+FROM python:3.9-alpine as installer
+COPY --link tests/projects/job/ /tests/projects/job/
+WORKDIR /tests/projects/job
+
+FROM installer as builder
+COPY tests/projects/job/src/hello-world.py helloworld.py
```
Comment (SamTheisens) [inline tests/projects/ephemeral/deployment/Dockerfile-mpl:7]: It's a bit strange to me that stuff from job is pulled in here.

**Review by SamTheisens — CHANGES_REQUESTED** [review]
Looks good 👍 . Main comment regards use of the docker client.
I noticed there are no modifications to the documentation. Did you use it? If yes, was it clear?

File: src/mpyl/utilities/docker/__init__.py [hunk 6]
```diff
@@ -92,3 +92,21 @@ def build(logger: Logger, root_path: str, file_path: str, image_tag: str, target
...
+    if project.deployment is None:
+        raise RuntimeError(f'No deployment information was found for project: {project.name}')
```
Comment (SamTheisens) [inline src/mpyl/utilities/docker/__init__.py:99]: I think a `KeyError` is more appropriate here.

File: src/mpyl/steps/deploy/ephemeral_docker_deploy.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+            version='0.0.1',
+            stage=Stage.DEPLOY
+        ), produced_artifact=ArtifactType.NONE, required_artifact=ArtifactType.DOCKER_IMAGE)
+
+    def execute(self, step_input: Input) -> Output:
+        env_file_name = write_env_to_file(step_input.project, step_input.run_properties)
+        docker_image_name = docker_image_tag(step_input)
+
+        return custom_check_output(self._logger,
+                                   f"docker run --rm --env-file {env_file_name} {docker_image_name}")
```
[the remark this answers is not in the document]
Comment (SamTheisens) [reply src/mpyl/steps/deploy/ephemeral_docker_deploy.py:27]: You can use:
`docker.run(image=docker_image_name, remove=True, envs={'ENV_VAR', 'value'})`

File: src/mpyl/utilities/docker/__init__.py (hunk 2 above)
Comment (SamTheisens) [reply src/mpyl/utilities/docker/__init__.py:97]: @koeves I would suggest not to add any logic to `mpyl.project.Properties`, but rather keep it a dumb data container. 
Since `write_env_to_file` is only used in one place, I would suggest to keep it local (to `ephemeraldockerdeploy.py`).

Assuming we need it at all, that is. https://github.com/Vandebron/mpyl/pull/65#discussion_r1149964464

File: tests/projects/ephemeral/deployment/Dockerfile-mpl (hunk 5 above)
Comment (Jorg88) [reply tests/projects/ephemeral/deployment/Dockerfile-mpl:7]: Fair, you suggest we have exactly the same `src` folder in this project as well? Or do you have an entirely different idea on how to check the container is running?

Comment (koeves) [reply tests/projects/ephemeral/deployment/Dockerfile-mpl:7]: Maybe you can just do a `printenv` in the container to check if the env vars are there

File: src/mpyl/utilities/docker/__init__.py [hunk 8]
```diff
@@ -92,3 +92,17 @@ def build(logger: Logger, root_path: str, file_path: str, image_tag: str, target
...
+def get_env_variables(project: Project, run_properties: RunProperties) -> dict[str, str]:
+    if project.deployment is None:
+        raise KeyError(f'No deployment information was found for project: {project.name}')
+    if len(project.deployment.properties.env) == 0:
+        raise KeyError(f'No properties.env is defined for project: {project.name}')
+
+    env_variables: dict[str, str] = {}
+
+    for env_variable in project.deployment.properties.env:
+        env_variables[env_variable.key] = env_variable.get_value(run_properties.target)
```
Comment (koeves) [inline src/mpyl/utilities/docker/__init__.py:106]:
```suggestion
    env_variables: dict[str, str] = {
        env_variable.key: env_variable.get_value(run_properties.target) for env_variable in
        project.deployment.properties.env
    }
```
is more pythonic I think, though maybe not more readable..

File: tests/test_resources/mpyl_config.yml [hunk 9]
```diff
@@ -18,6 +18,7 @@ jira:
   userName: !ENV ${JIRA_USER_PASSWORD_USR:jira_user}
   password: !ENV ${JIRA_USER_PASSWORD_PSW:jira_password}
 docker:
+  envFileName: '.env'
```
Comment (koeves) [inline tests/test_resources/mpyl_config.yml:21]: Remove this

File: src/mpyl/schema/mpyl_config.schema.yml [hunk 10]
```diff
@@ -110,6 +110,9 @@ definitions:
     type: object
     additionalProperties: false
     properties:
+      envFileName:
+        description: "Name of the .env file required for ephemeral docker step"
+        type: string
```
Comment (koeves) [inline src/mpyl/schema/mpyl_config.schema.yml:115]: Also not this

File: mpyl_config.yml [hunk 11]
```diff
@@ -18,6 +18,7 @@ jira:
   userName: !ENV ${JIRA_USER_PASSWORD_USR:jira_user}
   password: !ENV ${JIRA_USER_PASSWORD_PSW:jira_password}
 docker:
+  envFileName: '.env'
```
Comment (koeves) [inline mpyl_config.yml:21]: Not needed anymore I think

File: tests/test_resources/mpyl_config.yml (hunk 9 above)
Comment (Jorg88) [reply tests/test_resources/mpyl_config.yml:21]: Ugh, I already [did](https://github.com/Vandebron/mpyl/pull/65/commits/e035025d6903b8c42159deae21b38ee6487cfa8a#diff-5c9c3ecafe366913e1c62262e90a186ed331e486791120777125d8bce648f7e4) so but all sorts of weird stuff hapened when I did a pull..

File: src/mpyl/utilities/docker/__init__.py (hunk 8 above)
Comment (Jorg88) [reply src/mpyl/utilities/docker/__init__.py]: Fine by me, although I have to say that for someone without a lot of Python experience the first method was easier

File: src/mpyl/utilities/docker/__init__.py [hunk 12]
```diff
@@ -100,9 +100,9 @@ def get_env_variables(project: Project, run_properties: RunProperties) -> dict[s
...
         raise KeyError(f'No properties.env is defined for project: {project.name}')
 
-    env_variables: dict[str, str] = {}
-
-    for env_variable in project.deployment.properties.env:
-        env_variables[env_variable.key] = env_variable.get_value(run_properties.target)
+    env_variables: dict[str, str] = {
+        env_variable.key: env_variable.get_value(run_properties.target) for env_variable in
+        project.deployment.properties.env
+    }
```
Comment (SamTheisens) [inline src/mpyl/utilities/docker/__init__.py:106]: This is not related to docker and only used in one place. So better to pull local to where it's used I think.

File: src/mpyl/steps/deploy/ephemeral_docker_deploy.py [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+from ...project import Stage
+from ...utilities.docker import docker_image_tag, get_env_variables
+
+
+class EphemeralDockerDeploy(Step):
+
+    def __init__(self, logger: Logger) -> None:
+        super().__init__(logger, Meta(
+            name='Ephemeral Docker Deploy',
+            description='Deploy From Docker Container',
```
Comment (SamTheisens) [inline src/mpyl/steps/deploy/ephemeral_docker_deploy.py:18]:
```suggestion
            name='Deploy From Docker Container',
            description='Runs and removes the docker container built during the build stage. Useful for custom deploy steps depending on technology not bundled with MPyL',
```

File: src/mpyl/utilities/docker/__init__.py [hunk 14]
```diff
@@ -92,3 +92,17 @@ def build(logger: Logger, root_path: str, file_path: str, image_tag: str, target
     stream_docker_logging(logger=logger, generator=logs, task_name=f'Build {file_path}:{target}')
     logger.debug(logs)
     return True
+
+
+def get_env_variables(project: Project, run_properties: RunProperties) -> dict[str, str]:
```
Comment (SamTheisens) [inline src/mpyl/utilities/docker/__init__.py:97]: I think it would be cleaner to narrow the interfaces of the method arguments: pass in the list of properties and the `Target` only.

File: tests/projects/ephemeral/deployment/project.yml [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+        acceptance: "publicApiFacade.vdb-services.svc.cluster.local"
+        production: "publicApiFacade.vdb-services.svc.cluster.local"
+      - key: KEYCLOAK_HOST
+        pr: "keycloak.svc.cluster.local"
+        test: "keycloak.keycloak.svc.cluster.local"
+        acceptance: "keycloak.keycloak.svc.cluster.local"
+        production: "keycloak.keycloak.svc.cluster.local"
+    vault:
+      - key: ADMIN_API_KEY
+        path: "kong/gateway/admin_api_key"
```
Comment (SamTheisens) [inline tests/projects/ephemeral/deployment/project.yml:30]: Vault is dead

**Review by SamTheisens — CHANGES_REQUESTED** [review]
A few nitpicks

File: src/mpyl/utilities/docker/__init__.py (hunk 14 above)
Comment (Jorg88) [reply src/mpyl/utilities/docker/__init__.py:97]: True, we were using more out of `RunProperties` in the old structure

**Review by koeves — APPROVED** [review]

**Review by SamTheisens — APPROVED** [review]
